src/EventListener/UserSessionListener.php

```diff
@@ -0,0 +1,287 @@
+<?php
+
+/**
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2015 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\BackendUser;
+use Contao\CoreBundle\Session\Attribute\AttributeBagAdapter;
+use Contao\FrontendUser;
+use Doctrine\DBAL\Driver\Connection;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Attribute\AttributeBagInterface;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+/**
+ * Stores the session data of back end and front end users and restores it on
+ * the next login so the last application state can be restored.
+ *
+ * @author Yanick Witschi <https://github.com/toflar>
+ */
+class UserSessionListener extends ScopeAwareListener
+{
+    /**
+     * @var SessionInterface
+     */
+    private $session;
+
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * Constructor.
+     *
+     * @param SessionInterface $session
+     */
+    public function __construct(
+        SessionInterface $session,
+        Connection $connection
+    ) {
+        $this->session = $session;
+        $this->connection = $connection;
+    }
+
+    /**
+     * Checks whether the user has session data stored and passes it to the
+     * session service if so.
+     *
+     * @param GetResponseEvent $event The event object
+     */
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+        if (!$this->isFrontendMasterRequest($event)
+            && !$this->isBackendMasterRequest($event)
+        ) {
+            return;
+        }
+
+        $bag     = $this->getSessionBag();
+        $session = $this->getUserObject()->session;
+
+        // Restore session
+        if (is_array($session)) {
+            $bag->replace($session);
+        }
+    }
+
+    /**
+     * Checks whether there is something to store in the user session depending
+     * on the container scope.
+     *
+     * @param FilterResponseEvent $event The event object
+     */
+    public function onKernelResponse(FilterResponseEvent $event)
+    {
+        if (!$this->isFrontendMasterRequest($event)
+            && !$this->isBackendMasterRequest($event)
+        ) {
+            return;
+        }
+
+        $request = $event->getRequest();
+
+        if ($this->isBackendScope()) {
+            $this->storeBackendSession($request);
+            return;
+        }
+
+        $this->storeFrontendSession($request);
+    }
+
+    /**
+     * Stores the session for the back end.
+     *
+     * @param Request $request
+     */
+    private function storeBackendSession(Request $request)
+    {
+        if (!$this->canModifyBackendSession($request)) {
+            $this->storeSession();
```

Comment:
Because you always need to store the data of the session? We only modify the referer part but there are a lot of other things going on. This is also done in Contao 3 (storing is outside the `if` clause). Also see @ausi's comment here: https://github.com/contao/core-bundle/pull/207#issuecomment-93442689